compiler/src/dotty/tools/dotc/transform/BetaReduce.scala

```diff
@@ -48,17 +51,22 @@ object BetaReduce:
   import ast.tpd._
 
   /** Beta-reduces a call to `fn` with arguments `argSyms` or returns `tree` */
-  def apply(tree: Apply, fn: Tree, args: List[Tree])(using Context): Tree =
+  def tryReduce(fn: Tree, args: List[Tree])(using Context): Option[Tree] =
```

Comment:
Introducing an Option is a bit expensive, what's wrong with returning the input if there was nothing to reduce?

Reply:
It was just tricky to get right. I got a couple of bugs while trying that version an the realized that it would be much simpler to do it this way.

Reply:
Though, now that I encoded it this way it is simpler to optimize out the Option.